### DISCUSSION TITLE
Bad timing

### DISCUSSION THREAD

**Discussion by kybos** [muse-sequencer/muse#1207] [Team Posts]
Hi guys,

Incidentally I was able to reproduce the problem a user complained about in the forum recently:
![image](https://user-images.githubusercontent.com/10009541/111900663-29d9bc80-8a34-11eb-9a66-d04e4d559b24.png)
The link to the old homepage must be of course changed, but where is this FAQ that describes the remedy?

**Comment by terminator356** [comment]
I found the forum post.
I have been meaning to pipe up and discuss and explain this for a long time now.
Let me explain it here first before replying to the forum.

First, no remedy is described in any FAQ because this is the first time I have talked about this.

Second, it's a giveaway that you are running the dummy driver, or possibly the RT Pulse or some other driver,
 but **not** the Jack driver. Here's why:

I have researched the problem. It appears to be an ALSA problem.
We simply ask ALSA for the available timers and their resolutions, and that is what we get.

Try this experiment, to better understand what is happening:
1) Start your computer but **do not** start Jack at all.
2) Start MusE with the dummy driver by using the -a switch.
3) Observe the above bad timing warning. The ALSA timer only has 250Hz resolution at this point !
4) Close MusE.
5) Start the Jack server, say with QJackCtl or command line.
6) Start MusE **again with the dummy driver** by using the -a switch.
7) Observe that the **warning is gone** and that the timer now has a resolution of 1000Hz (or better).

Now here's the eye-opening part:

8) Close MusE again.
9) Stop the Jack server, say by Stop in QJackCtl and close QJackCtl, or by terminating the command line.
10) Start MusE **again with the dummy driver** by using the -a switch.
11) Observe that the **warning is STILL gone** and that the timer **still** has a resolution >=1000Hz.

You see?
Before Jack is started, the ALSA timers have a resolution of 250Hz.
After Jack is run and **even after it is stopped**, the ALSA timers have a resolution of 1000Hz.

Therefore I concluded that likely the mere act of Jack (or any other ALSA audio application) 
 opening the ALSA audio interface does something to the ALSA timers to "bump up" or "kick"
 their maximum resolution higher.

Which means that to solve this, MusE should probably try opening and then closing
 the ALSA audio interface at startup - even though currently we would never actually use it 
 (unless we bring back our very old pure ALSA audio driver, worth investigating, Ardour did that) !

So far I think I could not find some ALSA function call that would do that without opening/closing
 the audio interface. After all, we simply query the timers, and that's what they tell us.

**Comment by terminator356** [comment]
Forgot to mention.
So the 'remedy' is to always run Jack before starting MusE - or at least ensure that it runs once
 even if it is just closed again.

**Comment by spamatica** [comment]
There used to be a FAQ item about reconfiguring timers. As far as I remember a linux system has at least RTC, alsa-timer and HPET. We support alsa and RTC, not sure about HPET. The RTC timer is (or used to be) reconfigurable to use a higher frequencies.

Regarding the ALSA timer, it's frequency can be read from /proc/asound/timers. That something reconfigures it is curious, I didn't know that was possible. Maybe alsa is configured to use RTC and _that_ is reconfigured... just a guess. There is a file /proc/devices/rtc for it's settings so maybe they can be compared. 
I think there is a file somewhere under /sys for setting RTC.

The alert dialog about less than 500hz, as I recall I added that as MusE is quite dependent on the timer frequency (I might be mistaken but I think MusE might be the last "big" sequencer in Linux that has it's own timer handling in this way, others rely on alsa or jack). 
The check for 500hz (and the text that 250hz should be OK) was a somewhat educated guess as to what timing is good enough not to be audible.

**Comment by spamatica** [comment]
All hail the wayback machine, found a copy of the FAQ
http://web.archive.org/web/20190503064114/http://www.muse-sequencer.org/index.php/Faq

**Comment by terminator356** [comment]
**Update:** Now I recall some more.

The real problem is that snd_hrtimer is **not loaded** at system start. Type lsmod and observe.
The 'remedy' is to type, as root, "modprobe snd_hrtimer".

What I found was that Jack, or any other ALSA audio app, by virtue of opening the ALSA audio
 interface, automatically loads snd_hrtimer.

And my problem was that I couldn't find a way to tell the system to automatically load
 snd_hrtimer by loading all the other usual modules such as seq, timer, snd etc.
It's just not brought into the discussion by default.

So it seems like it could be an ALSA bug.
I would think that if I am opening snd_seq and snd_timer and so on, snd_hrtimer should
 by all reasonable consideration automatically load as well.

**Comment by terminator356** [comment]
I recall researching whether an app can tell the system to load a kernel module.
I think there was a way but I think maybe I couldn't get it to work.
If we can make that work, we can have MusE tell the system to load the module
 to make sure it's always loaded...

**Comment by terminator356** [comment]
Oh yeah, perhaps permissions got in the way!
MusE might need root permissions to load the module. Or not? Can't recall the method used...

**Comment by spamatica** [comment]
Yeah, loading a module is no problem if we are root but that seems way outside our jurisdiction. 
I think the current warning is a good way of handling it, we should however verify that the fixes suggested in the old faq still are valid and republish them. With the addition of your findings Tim, that starting jack (or other suitable application) ~should~/could clear it up.

**Comment by terminator356** [comment]
You might have missed my above comment. I expect a new comment soon, he he...
As stated above, the 'remedy' is to type, as root, "modprobe snd_hrtimer".

**Comment by terminator356** [comment]
Eek, The ol' memory is fading. Here is a post from two years ago, explaining what to do.
How about we put that info in the README or somewhere?
[https://linuxmusicians.com/viewtopic.php?t=19896](url)

**Comment by spamatica** [comment]
Not sure how you pasted the link, it doesn't work. No matter, copy paste works :)
Good background but we should give some hands on examples also.

And yeah, I missed the bit about modprobe. If that solves it, it seems like a good solution! Should be possible to add to some /etc file to default load this module upon boot.

**Comment by spamatica** [comment]
On my system there is the file /etc/modules-load.d/modules.conf
I simply added the snd_hrtimer last in the list and it is successfully loaded on boot
As I currently don't have this problem I can't tell if it solves the problem but the mechanism seems sound.

**Comment by spamatica** [comment]
I tried my best to revive the FAQ, placed it under a new Support link as I couldn't figure out where else to put it. Kybos as you organised most of it, if you have a better idea, I can change it.
 https://github.com/muse-sequencer/muse/wiki/Support#enable-high-performance-timer-in-alsa

**Comment by kybos** [comment]
I think the timer related part could have its own section in Special topics.
As for the rest, it seems that there is partly outdated and misleading information in it so I'm not sure if it were not better hidden. But if you think it still useful then perhaps keep it as FAQ (instead of Support), then people would know what to expect.

**Comment by terminator356** [comment]
I just did some edits to that.
But still some outdated info in there.
The best place for the most recent info is the main README, some of which I copied to this page.

**Comment by spamatica** [comment]
Alright, cool. Regarding the precount features, I vote for removing the preroll and prerecord from the dialog, I don't know either what they are supposed to be.

Yeah, there is still some pretty useless info there. I will try to go through it again and maybe move the sensible parts as you say kybos.

**Comment by terminator356** [comment]
Go ahead and remove them, and that piece of prerecord/preroll text on the page.
[ Edit ] Actually you can remove that whole precount text section after that since it works fully.

I think that the features I gave it should be sufficient for now until somebody complains.
